Submariner collapse operator crds

mangelajo opened the pull request:
This pull request collapses the RouteAgent CRD of the operator into the Submariner CRD. The submariner controller now reconciles both the DaemonSet for the routeagent and the Deployment for the submariner engine.

mangelajo committed:
Move all reconcile logic to the Submariner operator CRD

When we check-in the operator code we can refactor the
submariner_controller.go in several files or modules, now
it's a monolithic.

mangelajo committed:
Remove the RouteAgent CRD from the operator.

This CRD is not necessary anymore, and everything is handled
from the Submariner CRD.

mangelajo commented:
Passed E2E correctly:

Ran 6 of 6 Specs in 64.707 seconds
SUCCESS! -- 6 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
ok  	github.com/submariner-io/submariner/test/e2e	64.714s
your 3 virtual clusters are deployed and working properly with your local
submariner source code, and can be accessed with:

```
[ajo@rhvirt-casa submariner]$ kubectl --context=cluster2 get DaemonSets --all-namespaces
NAMESPACE     NAME         DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
kube-system   kube-proxy   3         3         3       3            3           <none>          8m16s
kube-system   weave-net    3         3         3       3            1           <none>          7m43s
operators     routeagent   2         2         2       2            2           <none>          4m49s
[ajo@rhvirt-casa submariner]$ kubectl --context=cluster2 get Deployments --all-namespaces
NAMESPACE     NAME                  READY   UP-TO-DATE   AVAILABLE   AGE
default       netshoot              1/1     1            1           4m17s
kube-system   coredns               2/2     2            2           8m22s
operators     submariner            1/1     1            1           4m55s
operators     submariner-operator   1/1     1            1           4m58s
[ajo@rhvirt-casa submariner]$ kubectl --context=cluster2 get pods -n operators
NAME                                  READY   STATUS    RESTARTS   AGE
routeagent-8v5fk                      1/1     Running   0          5m4s
routeagent-px5q9                      1/1     Running   0          5m4s
submariner-554c9d4dfd-s2g45           1/1     Running   0          5m4s
submariner-operator-695955c7c-lvffs   1/1     Running   0          5m7s

```

dfarrell07 commented on operators/go/submariner_controller.go.nolint:
```diff
@@ -253,3 +295,74 @@ func newPodTemplateForCR(cr *submarinerv1alpha1.Submariner) corev1.PodTemplateSp
+					Containers: []corev1.Container{
+						{
+							Name:  "submariner-routeagent",
+							Image: "submariner-route-agent:local",
```
We need to add the new cr.Spec.Image field, as introduced by https://github.com/dfarrell07/submariner/pull/21

skitt replied:
And we’ll need to revisit that slightly to distinguish engine and route agent images.

mangelajo replied:
I have been looking at how operators (from operatorhub) approach this [1][2][3][4]. To start with, there is no standard.

The model I liked the most was etcd's, they provide two fields like this:
```
	// Repository is the name of the repository that hosts
	// etcd container images. It should be direct clone of the repository in official
	// release:
	//   https://github.com/coreos/etcd/releases
	// That means, it should have exact same tags and the same meaning for the tags.
	//
	// By default, it is `quay.io/coreos/etcd`.
	Repository string `json:"repository,omitempty"`

	// Version is the expected version of the etcd cluster.
	// The etcd-operator will eventually make the etcd cluster version
	// equal to the expected version.
	//
	// The version must follow the [semver]( http://semver.org) format, for example "3.2.13".
	// Only etcd released versions are supported: https://github.com/coreos/etcd/releases
	//
	// If version is not set, default is "3.2.13".
	Version string `json:"version,omitempty"`
```


In our case, I think repository could be something like 
```
Repository: 'quay.io/submariner-io'
# Version: 'master'
```
For example, on development we could use:
```
Repository: ''   
Version: 'local'
```


Somebody else could publish their own images (for their own purposes) on another org / personal account if they needed, that'd help experimentation:
```
Repository: 'docker.io/rancher'
```
or
```
Repository: 'quay.io/mangelajo'
```

etc...

I'm going to rebase on your latest patch, and apply this to see how it looks.



[1] https://github.com/radanalyticsio/spark-operator/blob/master/examples/cluster-with-config.yaml#L17
[2] https://github.com/eclipse/che-operator/blob/master/pkg/apis/org/v1/che_types.go#L35
[3] https://github.com/coreos/etcd-operator/blob/master/pkg/apis/etcd/v1beta2/cluster.go#L80
[4] https://github.com/rook/rook/blob/master/pkg/apis/ceph.rook.io/v1/types.go#L103

skitt replied:
I like the “Image” variable approach, since it gives a single value to fill in which handles all situations (other repos, local versions etc.), which makes the docs simpler ;-).

mangelajo replied:
Oh skitt, I didn't see your message, sorry

I think that using the "Image" instead of repo + version combination substracts advantages from the operator and adds more chances for human error (for example, mixing versions that don't work well together, etc..)

mangelajo replied:
On the other hand, having 2 image fields, gives you freedom to use one version of the image from somewhere, and another version of the image from devel, troubleshooting, etc.

I have seem some operators which add image override fields, to override any manual calculation, and force it, like

imageOverrides:
    engine: docker.io/myrepo/submariner
    # routeAgent:

We also need to think about how migrations will happen from A to B, for example, when we are migrating from 0.1 to 0.2, to 0.3 ... may be the operator needs to know that the target version is 0.3, but the current version (in status) is 0.1 , so it needs to execute migration steps as it progresses and migrates components in the right order, etc..

We get no simple topic ... 'XD

skitt replied:
OK, I wasn’t thinking of a single repo + version shared for both images, I was imagining four fields altogether (to replace the two image fields). Sharing the repo and version does make more sense by limiting the amount of rope we give the users! Migrations will need some work, yes; we’ll probably have to limit the versions then (and perhaps use a development tweak in the operator to allow us to override it).
Operator migration capabilities are described in the CSV in the package, so we get some more tweaks there; for example, we can limit the possible upgrade paths (but between versions of the operator, not versions of the images pulled by the operator).

dfarrell07 reviewed (changes requested):
As we rebase this on the latest of the operator branch one less-obvious conflict is noted inline.

mangelajo committed:
Modify the Image field, to become Repository and Version

In such way, the operator can infer the correct image names it
needs for running, the images can be local to the cluster (CI/devel),
or remote.

The default repository is 'quay.io/submariner', and the default version
now is 'latest', that would result in:

quay.io/submariner/submariner:latest
quay.io/submariner/submariner-route-agent:latest

The keyword 'local' can be used as a repository so it can be ommited
on the path to the images, for example repository=local and version=local
would result in:

submariner:local
submariner-route-agent:local

being used, and that's what we use on the e2e testing.

mangelajo commented:
Ok, and it works. The last changes related to the image conversation are here on the last commit:

https://github.com/dfarrell07/submariner/pull/31/commits/67087ef47fbb9f25d6351da5ddc9609f6958c84c

Ran 6 of 6 Specs in 64.551 seconds
SUCCESS! -- 6 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

mangelajo commented on operators/go/submariner_controller.go.nolint:
```diff
@@ -365,3 +365,40 @@ func newRouteAgentDaemonSet(cr *submarinerv1alpha1.Submariner) *appsv1.DaemonSet
+
+//TODO: move to a method on the API definitions, as the example shown by the etcd operator here :
+//      https://github.com/coreos/etcd-operator/blob/8347d27afa18b6c76d4a8bb85ad56a2e60927018/pkg/apis/etcd/v1beta2/cluster.go#L185
+func setSubmarinerDefaults(submariner *submarinerv1alpha1.Submariner) {
```
Oh, I forgot to apply the defaults... 1 sec ':)

mangelajo replied:
done

mangelajo commented on scripts/kind-e2e/lib_operator_verify_subm.sh:
```diff
@@ -322,7 +280,7 @@ function verify_subm_routeagent_pod() {
   for subm_routeagent_pod_name in "${subm_routeagent_pod_names_array[@]}"; do
     echo "Testing Submariner routeagent pod $subm_routeagent_pod_name"
     kubectl get pod $subm_routeagent_pod_name --namespace=$subm_ns -o json
-    kubectl get pod $subm_routeagent_pod_name --namespace=$subm_ns -o jsonpath='{.spec.containers..image}' | grep $subm_routeagent_image_repo:$subm_engine_image_tag
```
for the engine we also do submariner:$subm_engine_image_tag so I removed the variable and set it here too..

dfarrell07 reviewed (approved):
LGTM! Passes codegen/e2e/ci for me locally and looks good statically. Thanks @mangelajo, big/great change! :pray:

mangelajo commented:
wow thanks! :-)
